wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/ConversationTableViewDataSource.swift

```diff
@@ -117,34 +118,120 @@ final class ConversationTableViewDataSource: NSObject {
     ///
     /// - Parameter forceRecalculate: true if force recreate cell with context check
     /// - Returns: arraySection of cell descriptions
-    @discardableResult
     func calculateSections(
-        forceRecalculate: Bool = false
-    ) -> [Section] {
-        messages.enumerated().map { offset, element in
-            let sectionIdentifier = element.objectIdentifier
-            let context = context(
-                for: element,
-                at: offset,
-                firstUnreadMessage: firstUnreadMessage,
-                searchQueries: searchQueries
-            )
-            let sectionController = sectionController(for: element, at: offset)
-
-            // Re-create cell description if the context has changed (message has been moved around or received new
-            // neighbors).
-            if sectionController.context != context || forceRecalculate {
-                sectionController.recreateCellDescriptions(in: context)
+        forceRecalculate: Bool = false,
+        completion: @escaping ([Section]) -> Void
+    ) {
+        let messagesOnMainThread = allMessages
+        let messageIds = messagesOnMainThread.map(\.objectID)
+        let selfUserOnMainThread = userSession.selfUser
+        let selfUserObjectID = selfUserOnMainThread.objectId
+        let firstUnreadMessageNonce = firstUnreadMessage?.nonce
+
+        // Dispatching to background thread to offload sections calculation
+
+        let backgroundContext = userSession.contextProvider.viewBackgroundContext
+        backgroundContext.perform { [weak self] in
+            guard let self else { return }
+
+            var messages: [ZMMessage] = messageIds.compactMap { objectID in
+                try? backgroundContext.existingObject(with: objectID) as? ZMMessage
+            }
+
+            // sort if needed
+            messages = messages.sorted {
+                ($0.serverTimestamp ?? .distantPast) > ($1.serverTimestamp ?? .distantPast)
             }
 
-            return ArraySection(model: sectionIdentifier, elements: sectionController.tableViewCellDescriptions)
+            guard let selfUserOnBackgroundThread = getUserByIDUseCase.getUserByID(
```

Comment:
question: why have a use case to fetch the user from a different context and not use a static helper like we do for other entities?

```
let selfUserOnBackgroundThread = ZMUser.existingObject(id: selfUserObjectID, context: backgroundContext)
```